src/Adapter/Product/CustomizationFieldDeleter.php

```diff
@@ -0,0 +1,164 @@
+        $this->productProvider = $productProvider;
+        $this->productUpdater = $productUpdater;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function delete(CustomizationFieldId $customizationFieldId): void
+    {
+        $customizationField = $this->customizationFieldProvider->get($customizationFieldId);
+        $this->performDeletion($customizationField, CannotDeleteCustomizationFieldException::FAILED_DELETE);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function bulkDelete(array $customizationFieldIds): void
+    {
+        foreach ($customizationFieldIds as $customizationFieldId) {
+            $customizationField = $this->customizationFieldProvider->get($customizationFieldId);
+            $this->performDeletion($customizationField, CannotDeleteCustomizationFieldException::FAILED_BULK_DELETE);
+        }
+    }
+
+    /**
+     * @param CustomizationField $customizationField
+     * @param int $errorCode
+     *
+     * @return void
+     *
+     * @throws CannotDeleteCustomizationFieldException
+     * @throws CustomizationFieldException
+     * @throws ProductException
+     * @throws ProductNotFoundException
+     */
+    private function performDeletion(CustomizationField $customizationField, int $errorCode): void
+    {
+        $product = $this->getProduct((int) $customizationField->id_product);
+        $usedFieldIds = array_map('intval', $product->getUsedCustomizationFieldsIds());
+        $fieldId = (int) $customizationField->id;
+
+        try {
+            if (in_array($fieldId, $usedFieldIds)) {
+                $successfullyDeleted = $customizationField->softDelete();
+            } else {
+                $successfullyDeleted = $customizationField->delete();
+            }
+
+            if (!$successfullyDeleted) {
+                throw new CannotDeleteCustomizationFieldException(
+                    sprintf('Failed deleting customization field #%d', $customizationField->id),
+                    $errorCode
+                );
+            }
+
+            $this->productUpdater->refreshProductCustomizabilityFields($product);
+            $this->productUpdater->update($product, CannotUpdateProductException::FAILED_UPDATE_CUSTOMIZATION_FIELDS);
+        } catch (PrestaShopException $e) {
+            throw new CustomizationFieldException(
+                sprintf(
+                    'Error occurred when trying to delete customization field #%d',
+                    $fieldId
+                ),
+                0,
+                $e
+            );
+        }
+    }
+
+    /**
+     * @param int $productId
+     *
+     * @return Product
+     *
+     * @throws ProductException
+     * @throws ProductNotFoundException
+     */
+    private function getProduct(int $productId): Product
+    {
+        if (isset($this->productsById[$productId])) {
+            return $this->productsById[$productId];
+        }
+
+        return $this->productProvider->get(new ProductId($productId));
```

Comment:
```suggestion
        if (!isset($this->productsById[$productId])) {
            $this->productsById[$productId] = $this->productProvider->get(new ProductId($productId));
        }

        return $this->productsById[$productId];
```

Reply:
Or the cache field always remains empty ans is useless :)

Reply:
sure, nice catch :smile: